### PR TITLE
Add deferred diagnostics when AUX builtins appear in device code

### PR DESCRIPTION
**Pull request by srividya-sundaram**
If a sycl application has host code with target specific builtins and the device code is compiled for a different target, then the SYCL device compiler will compile both host and device source code and the resulting device build will have the SPIRV target triple, as well as the host’s target triple as an ‘aux-triple’   

We must make sure that this does not get codegen'ed. So we detect any AUX target builtins in device code and use the notion of deferred diagnostics where we are compiling for the device, but we don't know that this function will be codegen'ed for device yet. So we  create a diagnostic which is emitted if and when  the function will be codegen'ed.

### CODE REVIEW

File: clang/lib/Sema/SemaChecking.cpp [hunk 1]
```diff
@@ -1938,11 +1938,25 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
...
-        return ExprError();
+              Context.BuiltinInfo.getAuxBuiltinID(BuiltinID), TheCall)) {
+                /*At this point, we know that the builtin is an aux-builtin and also checked.
+                  AUX builtins are not allowed to appear inside device code.
+                  To handle this, we use the notion of "deferred diagnostics", where we are compiling
+                  for the device, but we don't know that this function will be codegen'ed
+                  for device yet. So we create a diagnostic which is emitted if and when we
+                  realize that the function will be codegen'ed
+                */
+                if (getLangOpts().SYCLIsDevice) {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaChecking.cpp]: This isn't right,tis will only diagnose when the builtin has some sort of error. (CheckTSBuiltinFunction call returns true on failure).  We STILL want to report those and let that be an expr error.  Otherwise we want to diagnose like youre doing below.

File: clang/lib/Sema/SemaChecking.cpp [hunk 2]
```diff
@@ -1938,11 +1938,25 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
...
+                    SYCLDiagIfDeviceCode(TheCall->getBeginLoc(), diag::err_aux_target_builtin_in_device_code);
+                    return ExprError();
```
Comment (erichkeane) [inline clang/lib/Sema/SemaChecking.cpp]: Don't return an expr-error in this branch, since this might end up being valid code (in the event that this is NOT device code).  You want to diagnose and let it continue.

File: clang/lib/Sema/SemaChecking.cpp (hunk 1 above)
Comment (srividya-sundaram) [reply clang/lib/Sema/SemaChecking.cpp]: I don't understand your comment. 
If `Context.BuiltinInfo.isAuxBuiltinID(BuiltinID)` is true, then the builtinID is an AUX builtin.
Then we call CheckTSBuiltinFunctionCall with the target architecture of the AUX builtin and the builtinID. If this is true, it means the AUX builtin does belong to the target specified. Right?
Then I check if we are inside device code or not, since we should only identify aux bultins inside device code. So I used - `if (getLangOpts().SYCLIsDevice)` which will be true, if we are inside device code?
>>(CheckTSBuiltinFunction call returns true on failure)
>>tis will only diagnose when the builtin has some sort of error.
Can you clarify what these 2 statements mean?

Comment (erichkeane) [reply clang/lib/Sema/SemaChecking.cpp]: > I don't understand your comment.
> If `Context.BuiltinInfo.isAuxBuiltinID(BuiltinID)` is true, then the builtinID is an AUX builtin.
> Then we call CheckTSBuiltinFunctionCall with the target architecture of the AUX builtin and the builtinID. If this is true, it means the AUX builtin does belong to the target specified. Right?

isAuxBuiltinID means that the target arch of the aux builtin is the aux target.  That part is already calculated.  The CheckTSBuiltinFunctionCall validates the parameters to said function call..  If it is 'true' this check fails. 

> Then I check if we are inside device code or not, since we should only identify aux bultins inside device code. So I used - `if (getLangOpts().SYCLIsDevice)` which will be true, if we are inside device code?
Correct.
> 
> > > (CheckTSBuiltinFunction call returns true on failure)
> > > tis will only diagnose when the builtin has some sort of error.
> > > Can you clarify what these 2 statements mean?

Since the function is checking the parameters to the builtin, it does 1 of 2 things:  Returns 'false' meaning there was no errors.  OR  Emits an error (like, bad parameter value), then returns true.

File: clang/lib/Sema/SemaChecking.cpp (hunk 2 above)
Comment (srividya-sundaram) [reply clang/lib/Sema/SemaChecking.cpp]: if `getLangOpts().SYCLIsDevice` returns true, doesn't it mean we are in device code?
Maybe I got that part wrong? 
If that's the case, then this if block will execute only for device code and its okay to return the error right?

Comment (erichkeane) [reply clang/lib/Sema/SemaChecking.cpp]: "True" means we are in device compilation, not necessarily in device  code.  It could be host code being compiled during the device compilation (which will be thrown out later).  However, we don't want to return ExprError in that case, since it defeats other checks later (when we might not be emitting an error!).

Comment (srividya-sundaram) [reply clang/lib/Sema/SemaChecking.cpp]: Ok. Makes sense now.

File: clang/lib/Sema/SemaChecking.cpp (hunk 1 above)
Comment (srividya-sundaram) [reply clang/lib/Sema/SemaChecking.cpp]: Usually, when bool functions return true, it means something worked correctly and false means something went wrong.
This reverse logic is confusing.

Comment (erichkeane) [reply clang/lib/Sema/SemaChecking.cpp]: The CFE typically has 'true' mean "had an error", its just the convention that was done when they started.

Comment (srividya-sundaram) [reply clang/lib/Sema/SemaChecking.cpp]: The function name could be `ValidateBuiltinFuncParams `and it could return true if no errors and false if there is some error.
Or keep the existing logic and change the func name to isInvalidBuiltinParams() or something.

Comment (erichkeane) [reply clang/lib/Sema/SemaChecking.cpp]: Thats just not the convention of the CFE.  Also, "Check" is a very common word for exactly this throughout the sema layer.  Doing it differently would be more confusing to those who are familiar with the code.

File: clang/lib/Sema/SemaChecking.cpp [hunk 3]
```diff
@@ -1938,11 +1938,18 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
...
-
+      
       if (CheckTSBuiltinFunctionCall(
               Context.getAuxTargetInfo()->getTriple().getArch(),
-              Context.BuiltinInfo.getAuxBuiltinID(BuiltinID), TheCall))
-        return ExprError();
+              Context.BuiltinInfo.getAuxBuiltinID(BuiltinID), TheCall)) {
+                    
+                    return ExprError();
+      } else {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaChecking.cpp]: Don't else after a return.

File: clang/lib/Sema/SemaChecking.cpp [hunk 4]
```diff
@@ -1938,11 +1938,18 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
     if (Context.BuiltinInfo.isAuxBuiltinID(BuiltinID)) {
       assert(Context.getAuxTargetInfo() &&
              "Aux Target Builtin, but not an aux target?");
-
+
```
Comment (erichkeane) [inline clang/lib/Sema/SemaChecking.cpp]: Unrelated change.

**Review by erichkeane — COMMENTED** [review]
Write a lit test for this, make sure it works in the cases as you expect, then upload this for review.
